### PR TITLE
feat(Compiler): Add file removal tracking in watch-run

### LINKED ISSUE #5072: Able to track file removal in `watch-run`
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the current behavior?**

It's not possible to get removed files using current `watching.getTimes()` API.

**What is the expected behavior?**

Provide some way to get the list of removed files from FS watcher.

**If this is a feature request, what is motivation or use case for changing the behavior?**

It will allow TypeScript (and other stateful loaders) to remove files from the inner state.

### PR DESCRIPTION
**Pull request by cacheflow**
**What kind of change does this PR introduce?**

- [x] Feature

**Did you add tests for your changes?**

- [x] Yes

**Does this PR introduce a breaking change?**

- [x] No

**What needs to be documented once your changes are merged?**

- Fixes #5072

That watchRun now exposes removed files.

### CODE REVIEW

File: lib/Compiler.js [hunk 1]
```diff
@@ -128,6 +128,7 @@ class Compiler extends Tapable {
 		/** @type {string|null} */
 		this.recordsOutputPath = null;
 		this.records = {};
+		this.removedFiles = [];
```
[the remark this answers is not in the document]
Comment (cacheflow) [reply lib/Compiler.js]: Done.

File: lib/node/NodeWatchFileSystem.js [hunk 2]
```diff
@@ -57,7 +57,8 @@ class NodeWatchFileSystem {
 				changes.filter(file => dirs.includes(file)).sort(),
 				changes.filter(file => missing.includes(file)).sort(),
 				times,
-				times
+				times,
+				removals
```
Comment (sokra) [inline lib/node/NodeWatchFileSystem.js]: filter it by `files`, `dirs` or `missing`, sort it, convert it to a `Set`

File: lib/node/NodeWatchFileSystem.js [hunk 3]
```diff
@@ -57,7 +57,8 @@ class NodeWatchFileSystem {
 				changes.filter(file => dirs.includes(file)).sort(),
 				changes.filter(file => missing.includes(file)).sort(),
```
Comment (sokra) [inline lib/node/NodeWatchFileSystem.js]: For performance reasons, you could convert `files`, `dirs`, `missing` into Sets.

And add a TODO to change the `watch(files, dirs, missing` signature to `Set`s in webpack 5.

File: lib/node/NodeWatchFileSystem.js (hunk 2 above)
Comment (cacheflow) [reply lib/node/NodeWatchFileSystem.js]: 👍

File: lib/node/NodeWatchFileSystem.js (hunk 3 above)
Comment (cacheflow) [reply lib/node/NodeWatchFileSystem.js]: I'll change them into sets and will also take the TODO to change the signature in Webpack 5.

File: lib/node/NodeWatchFileSystem.js (hunk 2 above)
Comment (cacheflow) [reply lib/node/NodeWatchFileSystem.js]: Updated @sokra 👍

**Comment by sokra** [comment]
Thanks

**Pull request merged**